textacy/fileio/read.py

```diff
@@ -0,0 +1,187 @@
+"""
+Module with functions for reading content from disk in common formats.
+"""
+import bz2
+from functools import partial
+import gzip
+import io
+import json
+import os
+
+import ijson
+from spacy.tokens.doc import Doc
+
+
+JSON_DECODER = json.JSONDecoder()
+
+
+def read_json(filename, mode='rt', encoding=None, prefix=''):
+    """
+    Iterate over JSON objects matching the field given by ``prefix``.
+    Useful for reading a large JSON array one item (with ``prefix='item'``)
+    or sub-item (``prefix='item.fieldname'``) at a time.
+
+    Args:
+        filename (str): /path/to/file on disk from which json items will be streamed,
+            such as items in a JSON array; for example::
+
+                [
+                    {"title": "Harrison Bergeron", "text": "The year was 2081, and everybody was finally equal."},
+                    {"title": "2BR02B", "text": "Everything was perfectly swell."}
+                ]
+
+        mode (str, optional)
+        encoding (str, optional)
+        prefix (str, optional): if '', the entire JSON object will be read in at once;
+            if 'item', each item in a top-level array will be read in successively;
+            if 'item.text', each array item's 'text' value will be read in successively
+
+    Yields:
+        next matching JSON object; could be a dict, list, int, float, str,
+            depending on the value of ``prefix``
+
+    Notes:
+        Refer to ``ijson`` at https://pypi.python.org/pypi/ijson/ for usage details.
+    """
+    with io.open(filename, mode=mode, encoding=encoding) as f:
+        if prefix == '':
+            yield json.load(f)
+        else:
+            for item in ijson.items(f, prefix):
+                yield item
+
+
+def read_json_lines(filename, mode='rt', encoding=None):
+    """
+    Iterate over a stream of JSON objects, where each line of file ``filename``
+    is a valid JSON object but no JSON object (e.g. array) exists at the top level.
+
+    Args:
+        filename (str): /path/to/file on disk from which json objects will be streamed,
+            where each line in the file must be its own json object; for example::
+
+                {"title": "Harrison Bergeron", "text": "The year was 2081, and everybody was finally equal."}\n
+                {"title": "2BR02B", "text": "Everything was perfectly swell."}
+
+        mode (str, optional)
+        encoding (str, optional)
+
+    Yields:
+        dict: next valid JSON object, converted to native Python equivalent
+    """
+    with io.open(filename, mode=mode, encoding=encoding) as f:
+        for line in f:
+            yield json.loads(line)
+
+
+def read_json_mash(filename, mode='rt', encoding=None, buffersize=2048):
+    """
+    Iterate over a stream of JSON objects, all of them mashed together, end-to-end,
+    on a single line of a file. Bad form, but still manageable.
```

Comment:
Yeah, I've seen it done. "JSON lines" format is a _much_ better non-valid but streaming-friendly format.